### PR TITLE
NDRS-453 - carry over finalized deploys from joiner to validator phase

### PR DESCRIPTION
**Pull request by dwerner**
This PR adds the `linear_chain` as an argument to the `DeployBuffer`'s constructor, allowing it to load the finalized deploys from storage.

### CODE REVIEW

File: node/src/components/deploy_buffer.rs [hunk 1]
```diff
@@ -116,10 +129,38 @@ impl DeployBuffer {
...
+                    .event(move |result| {
+                        let deploys = result
+                            .iter()
+                            .flatten()
+                            .map(|deploy| (*deploy.id(), deploy.header().clone()))
+                            .collect::<HashMap<_, _>>();
+                        Event::GetDeploysForFinalizedBlockResult {
+                            block_hash,
+                            deploys,
+                        }
```
Comment (goral09) [inline node/src/components/deploy_buffer.rs:159]: There's a danger that we can't ignore that those events will be handled too late, after consensus asks deploy buffer for deploys. Why not simply create a `finalized_deploys` from the `linear_chain` blocks? There's enough data to reconstruct it, right?

File: node/src/components/deploy_buffer.rs [hunk 2]
```diff
@@ -85,7 +84,7 @@ impl Display for Event {
 }
 
 type DeployCollection = HashMap<DeployHash, DeployHeader>;
-type ProtoBlockCollection = HashMap<ProtoBlockHash, DeployCollection>;
+pub type ProtoBlockCollection = HashMap<ProtoBlockHash, DeployCollection>;
```
Comment (goral09) [inline node/src/components/deploy_buffer.rs:87]: I don't like making it pub type but I understand the alternative will be a mouthful.

**Review by goral09 — APPROVED** [review]
This looks fine to me but let's wait for @Fraser999  to review it on Monday as well. We can still approve it before the sprint exit.

**Comment by dwerner** [comment]
@Fraser999 and I had a call where we discussed some changes (move method to storage, call `SmallVec::from`), and those are my last commit.

**Comment by goral09** [comment]
> @Fraser999 and I had a call where we discussed some changes (move method to storage, call `SmallVec::from`), and those are my last commit.

Yep, looks OK. Thanks.

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**